### ISSUE TITLE
Using batch training and learn-by-demo in act-experience-update

### ISSUE THREAD

**Issue by ZYtheNOOB** [tensorforce/tensorforce#756]
Hi,

I am using ppo agent and wish to use batch in `act-experience-update` interface to speed up training (it's slow and I still cannot use GPU). Is calling `agent.update()` once every few episodes equivalent to using batch training? Like this:
```
batch_counter = 0
for episode in range(100):
    ...
    agent.experience(...)
    if batch_counter % batch_size == 0:
        agent.update()
    batch_counter += 1
```

I'm also trying to use learn-by-demostration, is this a correct way of doing this (not sure about episode_internals)?
```
            states = environment.reset()
            internals = agent.initial_internals()
            terminal = False
            # randomly insert expert demostrations
            if random.random() > expert_ratio:
                while not terminal:
                    episode_states.append(states)
                    episode_internals.append(internals)
                    actions, internals = agent.act(states=states, internals=internals, independent=True)
                    episode_actions.append(actions)
                    states, terminal, reward = environment.execute(actions=actions)
                    episode_terminal.append(terminal)
                    episode_reward.append(reward)
            else:
                # demo
                while not terminal:
                    episode_states.append(states)
                    episode_internals.append(internals)
                    states, terminal, reward = environment.execute(actions=-1)
                    exp_act = states.get('exp_action')
                    episode_actions.append(exp_act)
                    episode_terminal.append(terminal)
                    episode_reward.append(reward)

            agent.experience(
                states=episode_states, internals=episode_internals, actions=episode_actions,
                terminal=episode_terminal, reward=episode_reward
            )
```

**Comment by ZYtheNOOB** [comment]
Another curious thing I noticed in my previous trials using `act-observe`: in evaluation mode, tracking tensor `action-value` always gives the same values over all timesteps and all episodes, but at the end of my training the policy entropy is not very low (about 0.08). What might be the reason of this?

**Comment by AlexKuhnle** [comment]
Hi @ZYtheNOOB,

Regarding your questions:

(a) Yes, that should more or less be the same. One thing to keep in mind is that independent-act/experience/update doesn't increment the timesteps/episodes counter, so scheduled parameters won't work correctly. Other than that, I think it should work the same.

(b) RNNs and internal states are incompatible with demonstrations. (What are the internals for the demonstrations? Not possible, e.g. in your example you always use the initial internals.) So assuming that you don't use an RNN, i.e. empty internals, you should be able to simply skip the internals argument everywhere. Other than that it looks alright, assuming that the environment returns the correct rewards for `actions=-1`.

(c) What is your agent configuration?

**Comment by ZYtheNOOB** [comment]
Thanks for the clarification for the first two.
In (c), my agent is like this:
```
    agent = PPOAgent(
        states=environment.states(),
        actions=environment.actions(),
        network=nn.network_spec,
        batch_size=8,
        memory=100000,
        learning_rate=5e-4,
        exploration=0.02,
        discount=0.99,
        entropy_regularization=0.0001,
        state_preprocessing=dict(type='linear_normalization'),
        max_episode_timesteps=max_timesteps,
        saver=dict(directory='./saver/', filename='PPO_model.ckpt', load=evaluate, frequency=100, max_checkpoints=20),
        summarizer=dict(directory='./record/', summaries=["loss", "reward", "entropy", "graph"], flush=10),
        config={'seed': 3}
    )
```
The number of actions is 4, and after training one action always has a fixed probability of 0.97.

**Comment by AlexKuhnle** [comment]
For clarification: the probability of 0.97 is always the same *after training*? Is it possible that, since you also set a seed, the agent outputs are fully deterministic, which is why you always get the same value? Note that a random seed is set for TensorFlow as well as NumPy and Python. Does this value change during training? What if you change something else in the setup, or simply remove the random seed?

**Comment by ZYtheNOOB** [comment]
The action probabilities are the same for every timestep in evaluation after training. I get different values for different training, so I think it's not because of the random seed. For example in one training, the value does change from initial [0.25, 0.25, 0.25, 0.25] to something like [0.005, 0.97, 0.02, 0.005], and is fixed for any state input in evaluation (maybe the policy becomes "absolutely deteministic"?).

**Comment by AlexKuhnle** [comment]
Does the same happen if you stop training very early, e.g. after 1 episode, or are the actions/probabilities more diverse then? If this is the case, then the policy apparently overfits to one fixed distribution.

**Comment by ZYtheNOOB** [comment]
The actions start from [0.25, 0.25, 0.25, 0.25], which is quite normal. How can I reduce overfitting by changing agent's parameters?

**Comment by AlexKuhnle** [comment]
`entropy_regularization` is a good starting point. But it's also worth investigating why the agent overfits to one action -- shouldn't that result in bad performance?

And regarding my last comment: What I'm wondering is whether the distribution differs between different states, or whether it is always the same regardless of the state, as you reported for after training.

**Comment by ZYtheNOOB** [comment]
Regarding distribution: Yes they differs a little bit over states (in the first 50 epidose)
```
timestep: 29
[0.250331   0.25067952 0.24916065 0.24982882]
timestep: 30
[0.2502426  0.25072148 0.24923153 0.24980442]
timestep: 31
[0.25040877 0.2506318  0.24914743 0.24981198]
timestep: 32
[0.25048906 0.25047067 0.24918914 0.24985111]
timestep: 33
[0.25030264 0.25064036 0.24917094 0.24988611]
timestep: 34
[0.25037727 0.25044498 0.2491304  0.25004745]
```

Performance: the design of my action space may have some problems - I'm working on selective hyper-heuristics and always using one heuristic is kind of a local optima.

**Comment by ZYtheNOOB** [comment]
Regarding distribution: Yes they differs a little bit over states (in the first 50 epidose)
```
timestep: 29
[0.250331   0.25067952 0.24916065 0.24982882]
timestep: 30
[0.2502426  0.25072148 0.24923153 0.24980442]
timestep: 31
[0.25040877 0.2506318  0.24914743 0.24981198]
timestep: 32
[0.25048906 0.25047067 0.24918914 0.24985111]
timestep: 33
[0.25030264 0.25064036 0.24917094 0.24988611]
timestep: 34
[0.25037727 0.25044498 0.2491304  0.25004745]
```

Performance: the design of my action space may have some problems - I'm working on selective hyper-heuristics and always using one heuristic is kind of a local optima.

**Comment by AlexKuhnle** [comment]
This could be the reason for the "always 0.97" problem. Certainly it doesn't seem to be a problem with the Tensorforce output being "wrong", since apparently it does change for different states, right?

**Comment by ZYtheNOOB** [comment]
ok, thanks for your explanations.
